### ISSUE TITLE
lr scheduler state not saved with DeepSpeed plugin

### ISSUE THREAD

**Issue by eelxpeng** [PyTorchLightning/pytorch-lightning#11188]
## 🐛 Bug

When using deepspeed plugin, the learning rate scheduler state is not saved. The key `lr_scheduler` in `mp_rank_00_model_states.pt` model file is None. The bug causes the resuming from checkpoint to behave unexpected, where it should continue from previous lr scheduler state but instead lr scheduler start from scratch. This issue happens for most pytorch_lightning release versions, but the issue does not happen for some commits, e.g. commit 86c033572da084fd85d4d4c2d637db18d2d431ec.

### To Reproduce

```
import os
import torch
from torch.utils.data import Dataset, DataLoader
import pytorch_lightning as pl
from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks import ModelCheckpoint
from pytorch_lightning.plugins import DeepSpeedPlugin
from deepspeed.ops.adam import FusedAdam
from transformers.optimization import AdamW, get_scheduler

class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def train_dataloader(self):
        return DataLoader(RandomDataset(32, 64), batch_size=2)
    
    def test_dataloader(self):
        return DataLoader(RandomDataset(32, 64), batch_size=2)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return loss

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)
        return loss

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)
        return loss

    def configure_optimizers(self):
        no_decay = ["bias"]
        params_decay = [
            p for n, p in self.named_parameters() if not any(nd in n for nd in no_decay)
        ]
        params_nodecay = [
            p for n, p in self.named_parameters() if any(nd in n for nd in no_decay)
        ]
        optim_groups = [
            {
                "params": params_decay,
                "weight_decay": 0.01,
            },
            {"params": params_nodecay, "weight_decay": 0.0},
        ]
        optimizer = torch.optim.SGD(optim_groups, lr=0.1)
        scheduler = get_scheduler("linear", optimizer, num_warmup_steps=20, num_training_steps=200)
        return ([optimizer], [{"scheduler": scheduler, "interval": "step", "frequency": 1}])

def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=4)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=4)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=4)

    model = BoringModel()
    checkpoint_callback = ModelCheckpoint(
        dirpath='tests4/checkpoints',
        save_last=True,
        every_n_train_steps=10,
    )
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        gpus=-1,
        limit_train_batches=1.0,
        limit_val_batches=1,
        num_sanity_val_steps=0,
        precision=16,
        accelerator='deepspeed',
        max_steps=100,
        log_every_n_steps=1,
        plugins=[DeepSpeedPlugin(stage=2)],
        weights_summary=None,
        callbacks=[checkpoint_callback, pl.callbacks.LearningRateMonitor(logging_interval="step")],
        #resume_from_checkpoint='tests4/checkpoints/epoch=3-step=99.ckpt',
    )
    trainer.fit(model)
    trainer.test(model)

    # test lr_scheduler state
    model_state_path = "tests4/checkpoints/epoch=3-step=99.ckpt/global_step100/mp_rank_00_model_states.pt"
    model_state = torch.load(model_state_path, map_location="cpu")
    print(model_state["lr_schedulers"])


if __name__ == '__main__':
    run()
```

Just run `python run.py`. For latest release version 1.5.6 and many other versions including current master branch, it produces None, meaning lr_scheduler state is not saved. For commit id 86c033572da084fd85d4d4c2d637db18d2d431ec, it produces the correct state:
```
{'base_lrs': [0.1, 0.1], 'last_epoch': 47, '_step_count': 48, 'verbose': False, '_get_lr_called_within_step': False, '_last_lr': [0.05888888888888889, 0.05888888888888889], 'lr_lambdas': [None, None]}
```

Do first run with the code exactly as above. Do second run by changing max_steps from 100 to 200, and uncomment `#resume_from_checkpoint` line. The two runs gives the tensorboard curves of resuming training. For pytorch lightning version 1.5.6 and many others, it produces wrong lr curve:
<img width="346" alt="Screen Shot 2021-12-20 at 5 19 20 PM" src="https://user-images.githubusercontent.com/14260983/146855467-04cb6aa7-0684-4fae-bab8-2e84c8ea003d.png">

However, the correct resuming curve (like the one produced with commit id 86c033572da084fd85d4d4c2d637db18d2d431ec) should be
<img width="346" alt="Screen Shot 2021-12-20 at 5 16 57 PM" src="https://user-images.githubusercontent.com/14260983/146855585-3128b578-03c6-49b1-b76b-0e1143c858a1.png">

### Expected behavior

The lr_scheduler state should be saved properly in order to resume from checkpoint correctly. Could you clarify if we should handle lr_scheduler state loading or saving by ourselves, or it should be properly handled by pytorch_lightning? If it should be handled by pytorch_lightning, do you know how to resolve the bug?

cc @tchaton @SeanNaren @awaelchli

**Comment by awaelchli** [comment]
Thanks for reporting this. There could be a bug when assigning the scheduler using the Lightning `configure_optimizers` hook. It could be that it works when setting it via the config.

**Comment by eelxpeng** [comment]
@awaelchli Thanks for the reply. Would you elaborate "It could be that it works when setting it via the config"? As temporary workaround, how to set via config?

**Comment by tridao** [comment]
I'm seeing the same bug with Deepspeed not restoring the lr scheduler.

**Comment by awaelchli** [comment]
Ok, I came to the bottom of this. Here is the origin of this bug:

https://github.com/PyTorchLightning/pytorch-lightning/blob/05ed9a201c24e08c2b4d3df4735296758ddcd6a5/pytorch_lightning/strategies/deepspeed.py#L481-L487

If the scheduler is defined by the LightningModule (not via json config file), we set `deepspeed_engine.lr_scheduler = None`. This is all fine until we want to save a checkpoint, where we have this logic:

https://github.com/PyTorchLightning/pytorch-lightning/blob/05ed9a201c24e08c2b4d3df4735296758ddcd6a5/pytorch_lightning/strategies/deepspeed.py#L742-L746

We exclude the lr_scheduler key from the dict because assume deepspeed.save_checkpoint will save the lr scheduler dict, but since we set `deepspeed_engine.lr_scheduler = None` this will skip it. 

So I think the fix is to modify the `save_checkpoint` and `load_checkpoint` to take care of the lr scheduler dict whenever the scheduler is NOT managed by deepspeed internally.

**Comment by tchaton** [comment]
Hey @awaelchli,

Would changing `_exclude_keys` to `["state_dict", "optimizer_states"]` solve the bug ?

**Comment by awaelchli** [comment]
Yes I looked into this already, already discussed with @SeanNaren he agrees

**Comment by awaelchli** [comment]
Hello @tridao @eelxpeng 
If you have the time, could you quickly check my fixes in this branch here #11322 if it works with your use case? I have checked with your repro-example, just wanted to let you have a look.

**Comment by eelxpeng** [comment]
Updated the repro-example since there is some issue with it.
@awaelchli I checked the your fix, and it works great! See the following curves obtained. Thank you very much!
<img width="345" alt="Screen Shot 2022-01-05 at 3 12 45 PM" src="https://user-images.githubusercontent.com/14260983/148303037-0dac3cb3-396a-493d-a56e-2e8d303dfbe1.png">

**Comment by awaelchli** [comment]
Thanks for checking, glad we were able to make progress here. I will open the PR for review.
